File: bindings/kt/src/test/kotlin/LocalKeyManagerTests.kt
package web5.sdk

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import web5.sdk.LocalKeyManager
import kotlin.math.sign

class LocalKeyManagerTests {
    @Test
    fun `can generate a Ed25519 key`() {
        val keyManager = LocalKeyManager.newInMemory()
        val keyAlias = keyManager.generatePrivateKey(Curve.ED25519, null)
        assertNotEquals(0, keyAlias.length)
    }

    @Test
    fun `can generate a secp256k1 key`() {
        val keyManager = LocalKeyManager.newInMemory()
        val keyAlias = keyManager.generatePrivateKey(Curve.SECP256K1, null)
        assertNotEquals(0, keyAlias.length)
    }

    @Test
    fun `can sign and verify`() {
        val keyManager = LocalKeyManager.newInMemory()
        val keyAlias = keyManager.generatePrivateKey(Curve.ED25519, null)
        assertNotEquals(0, keyAlias.length)

        val payload = "hello world".toByteArray().map { it.toUByte() }
        val signature = keyManager.sign(keyAlias, payload)
        assertNotEquals(0, signature.size)

        val publicKey = keyManager.getPublicKey(keyAlias)
        assertDoesNotThrow {
            publicKey.verify(payload, signature.map { it.toUByte() })
        }
    }

    @Test
    fun `can export and import keys`() {
        val keyManager = LocalKeyManager.newInMemory()
        keyManager.generatePrivateKey(Curve.SECP256K1, null)

        val exportedPrivateKeys = keyManager.exportPrivateKeys()
        assertEquals(1, exportedPrivateKeys.size)

        assertDoesNotThrow {
            keyManager.importPrivateKeys(exportedPrivateKeys)
        }
    }
}

File: binded/swift/Tests/LocalKeyManagerTests.swift
import XCTest
@testable import Web5

final class LocalKeyManagerTests: XCTestCase {
    func testCanGenerateEd25519Key() throws {
        let keyManager = LocalKeyManager.newInMemory()
        let keyAlias = try keyManager.generatePrivateKey(curve: Curve.ed25519, keyAlias: nil)
        XCTAssertNotEqual(keyAlias.count, 0)
    }
    
    func testCanGenerateSecp256k1Key() throws {
        let keyManager = LocalKeyManager.newInMemory()
        let keyAlias = try keyManager.generatePrivateKey(curve: Curve.secp256k1, keyAlias: nil)
        XCTAssertNotEqual(keyAlias.count, 0)
    }
    
    func testCanSignAndVerify() throws {
        let keyManager = LocalKeyManager.newInMemory()
        let keyAlias = try keyManager.generatePrivateKey(curve: Curve.ed25519, keyAlias: nil)
        XCTAssertNotEqual(keyAlias.count, 0)
        
        let payload = "hello world".data(using: .utf8)!.map { $0 }
        let signature = try keyManager.sign(keyAlias: keyAlias, payload: payload)
        XCTAssertNotEqual(signature.count, 0)
        
        let publicKey = try keyManager.getPublicKey(keyAlias: keyAlias)
        XCTAssertNoThrow(try publicKey.verify(payload: payload, signature: Array(signature)))
    }
    
    func testCanExportAndImportKeys() throws {
        let keyManager = LocalKeyManager.newInMemory()
        _ = try keyManager.generatePrivateKey(curve: Curve.secp256k1, keyAlias: nil)
        
        let exportedPrivateKeys = try keyManager.exportPrivateKeys()
        XCTAssertEqual(exportedPrivateKeys.count, 1)
        
        XCTAssertNoThrow(try keyManager.importPrivateKeys(privateKeys: exportedPrivateKeys))
    }
}